astropy/table/tests/test_init_table.py

```diff
@@ -663,3 +663,11 @@ def test_init_QTable_and_set_units():
     assert np.all(t["col0"].value == [1000, 2000])
     assert t["col1"].unit == u.s
     assert np.all(t["col1"].value == [1, 2])
+
+
+def test_QTable_from_mixin_columns_with_exotic_str_type_name():
+    # see https://github.com/astropy/astropy/issues/17418
+    t = QTable()
+    t[np.str_("col1")] = [1, 2, 3]
+    t[np.str_("col2")] = [1, 2, 3] * u.m
+    assert [type(n) for n in t.colnames] == [str] * len(t.colnames)
```

Comment:
note to reviewers: if you're wondering if this breaks accessing a column as `t["col1"]`, the answer is no; because `Table.__getitem__` delegates to `OrderedDict.__getitem__` when the `item` is a `str`, and `hash(item) == hash(np.str_(item))`.